### ISSUE TITLE
ENGIN-X / IMAT: Instron (Stress Rig): IOC - Waveform Commands

### ISSUE THREAD

**Issue by KathrynBaker** [ISISComputingGroup/IBEX#2113]
As an IBEX user I want to be able to control the Instron Stress Rig natively in EPICS in order to leverage the best use of the system.

**Acceptance Criteria**
There are additions to the protocol and database files used by the IOC created in Ticket #2111 which covers the interactions for waveform control

**General Notes**
This ticket is of medium priority in relation to the Instron, ideally it would be in place for the cycle starting May 2nd 2017, however, an incomplete lvDCOM solution as specified in Ticket #2110 means that this can be delayed if required. This ticket should be done after Ticket #2111.

The rig in use consists of the channels, Position (POS) on Channel 1, Stress (STRESS) on Channel 2, ans Strain (STRAIN) on Channel 3.

There is a lot of information to support the stress rig work in a folder called Instron Stress Rig in the ICP discussions area of our local SharePoint. Items below are highlighted as being relevant to this ticket directly, some of the other information may be required from the other documents.

The PV names may not have included a $(P) or the correct number and position of `:`, and may not be the final names used, as clarity here was chosen over following the guidelines. Please use appropriate names, and if different from those in the documents - update the documents as you go

**Specific Notes**
PVs.xslx contains the list of PVs, and an indication of their interactions, the only entries to undertake as part of this ticket are those in the waveform commands section.

Some clarity, and more information can be found in Working through the block diagram.docx.

The Command reference.docx file has the subset of commands that are used by the existing LabVIEW driver, Stress Rig - GPIB.pdf has the full list of commands.

**Comment by kjwoodsISIS** [comment]
Need to do #2111 first.

**Comment by John-Holt-Tessella** [comment]
Don't review until #2112 is done

**Comment by Tom-Willemsen** [comment]
As just discussed, some tweaking to the waveform settings screen:

- [ ] There are currently 3 types of setpoint/readback pairs on this screen which each work in different ways. To solve this, reconvert the "amplitude" readback to be the same as what the user put in, and add the "raw value" separately.

- [ ] Amplitude readback should probably be in `mm` not `m` and the setpoint should have a unit.

- [ ] Add start/stop waveform buttons to waveform settings screen of OPI.

- [ ] A few minor comments and a couple of failing tests, see test framework PR

**Comment by John-Holt-Tessella** [comment]
Just two tests to fix:

1. In dev sim
```
ERROR [11.185s]: test_WHEN_the_quarter_counter_is_armed_THEN_the_number_of_quarts_never_exceeds_the_requested_maximum (tests.instron_stress_rig.Instron_stress_rigTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Instrument\Apps\EPICS\support\IocTestFramework\master\tests\instron_stress_rig.py", line 541, in test_WHEN_the_quarter_counter_is_armed_THEN_the_number_of_quarts_never_exceeds_the_requested_maximum
    self.assertLessEqual(4*float(self.ca.get_pv_value("QUART")), cycles)
AssertionError: 340.0 not less than or equal to 5
```

2. In rec sim

```
ERROR [0.002s]: test_GIVEN_the_waveform_generator_is_running_WHEN_instructed_to_hold_THEN_status_is_holding (tests.instron_stress_rig.Instron_stress_rigTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Instrument\Apps\EPICS\support\IocTestFramework\master\tests\instron_stress_rig.py", line 550, in test_GIVEN_the_waveform_generator_is_running_WHEN_instructed_to_hold_THEN_status_is_holding
    self._lewis.backdoor_command(["device", "set_waveform_state", WAVEFORM_RUNNING])
AttributeError: 'LewisNone' object has no attribute 'backdoor_command'
```

**Comment by AdrianPotter** [comment]
![image](https://user-images.githubusercontent.com/10273443/28966309-58a16ea0-790d-11e7-846d-8a89793fdf24.png)

**Comment by AdrianPotter** [comment]
I've done some rework and merged in all the changes from #2112. They all work for me now (rerun to check for stability). I hope they do for you too :)

**Comment by Tom-Willemsen** [comment]
Tests passed and code looks ok.

However I am getting the following messages in the emulator when setting the number of quarter cycles (and also when the emulator first starts up):

```
2017-08-07 13:49:27,239 INFO lewis.DeviceBase: Creating device, setting up state machine
2017-08-07 13:49:27,243 INFO lewis.Simulation: Changed cycle delay to 0.1
2017-08-07 13:49:27,244 INFO lewis.Simulation: Changed speed to 1.0
2017-08-07 13:49:27,246 INFO lewis.Simulation: Starting simulation
2017-08-07 13:49:27,253 INFO lewis.InstronStreamInterface.StreamServer: Listening on localhost:57677
2017-08-07 13:49:31,178 INFO lewis.InstronStreamInterface.StreamHandler: Client connected from 127.0.0.1:63330
2017-08-07 13:49:31,440 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q300 using command ^Q300$
2017-08-07 13:49:31,950 INFO lewis.InstronStreamInterface.StreamHandler: Processing request P909,1 using command ^P909,([0-1])$
Watchdog time expired, going back to front panel control mode
An error occurred at request 'C209,0': RuntimeError("None of the device's commands matched.",)
2017-08-07 13:49:31,956 INFO lewis.InstronStreamInterface.StreamHandler: Processing request P909,0 using command ^P909,([0-1])$
```

For command `C209` this exists in your protocol file but not in the emulator. Additionally, the emulator has a `Q209` command assigned to `set_max_quarter_counts` - this doesn't look right I think it should be `C209` here.

I also get the following shortly after starting up the emulator:

```
2017-08-07 13:49:32,467 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q86,2 using command ^Q86,([1-3])$
An error occurred at request 'Q201,0': RuntimeError("None of the device's commands matched.",)
2017-08-07 13:49:32,470 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q4,1 using command ^Q4,([1-3])$
2017-08-07 13:49:32,473 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q2,2 using command ^Q2,([1-3])$
2017-08-07 13:49:32,474 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q86,3 using command ^Q86,([1-3])$
An error occurred at request 'Q203,0': RuntimeError("None of the device's commands matched.",)
2017-08-07 13:49:32,480 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q134,1,0 using command ^Q134,([1-3]),([0-9]+)$
2017-08-07 13:49:32,484 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q4,2 using command ^Q4,([1-3])$
2017-08-07 13:49:32,486 INFO lewis.InstronStreamInterface.StreamHandler: Processing request Q2,3 using command ^Q2,([1-3])$
An error occurred at request 'Q202,0': RuntimeError("None of the device's commands matched.",)
```

Please could you check why the IOC is sending command arguments which the emulator doesn't expect? Currently the emulator expects the argument to be `[1-3]` From the manual, `Q201,0` for example means "next channel", but I don't think this is what you meant for the IOC to send?

Once you have fixed these could you add tests (sorry) to test that they behave as expected.

Please also add release notes.
